stan/math/prim/arr/functor/mpi_cluster.hpp

```diff
@@ -0,0 +1,251 @@
+ * interface with MPI. In boost mpi terminology we refer to the
+ * overall MPI ressources as world. Each process has an assigned
+ * rank which is a numeric index running from 0 to the number of
+ * processes-1 (also called the world size). We refer to the rank
+ * = 0 process as the root process and all others (rank > 0) are
+ * workers.
+ *
+ */
+struct mpi_cluster {
+  boost::mpi::environment env;
+  boost::mpi::communicator world_;
+  std::size_t const rank_ = world_.rank();
+  static std::mutex in_use_;
+
+  mpi_cluster() {}
+
+  ~mpi_cluster() {
+    // the destructor will ensure that the childs are being
+    // shutdown
+    stop_listen();
+  }
+
+  /**
+   * Switches cluster into listening mode. That is, for the root
+   * process we only flag that the workers are now listening and
+   * as such can recieve commands while for the non-root processes
+   * we enter into a listening state. In the listening state on
+   * the non-root processes we wait for broadcasts of mpi_command
+   * objects which are initiated on the root using the
+   * mpi_broadcast_command function below. Each recieved
+   * mpi_command is executed using the virtual run method.
+   */
+  void listen() {
+    mpi_cluster::cluster_listens_ = true;
+    if (rank_ == 0) {
+      return;
+    }
+    // TODO(wds15): Remove debugging print
+    std::cout << "Worker " << rank_ << " listening for commands..."
+              << std::endl;
+
+    try {
+      // lock on the workers the cluster as MPI commands must be
+      // initiated from the root and any attempt to do this on the
+      // workers must fail
+      std::unique_lock<std::mutex> worker_lock(in_use_);
+      while (1) {
+        boost::shared_ptr<mpi_command> work;
+
+        boost::mpi::broadcast(world_, work, 0);
+
+        work->run();
+      }
+    } catch (const mpi_stop_listen& e) {
+      // TODO(wds15): Remove debugging print
+      std::cout << "Wrapping up MPI on worker " << rank_ << " ..." << std::endl;
+    }
+  }
+
+  /**
+   * Stops listening state of the cluster. This happens
+   * automatically upon destruction. It is achieved by sending the
+   * mpi_stop_worker command.
+   */
+  void stop_listen() {
+    if (rank_ == 0 && mpi_cluster::cluster_listens_) {
+      mpi_broadcast_command<mpi_stop_worker>();
+    }
+    mpi_cluster::cluster_listens_ = false;
+  }
+
+  /**
+   * Returns the current listening state of the cluster.
+   */
+  static bool is_listening() { return cluster_listens_; }
+
+ private:
+  static bool cluster_listens_;
+};
+
+bool mpi_cluster::cluster_listens_ = false;
+std::mutex mpi_cluster::in_use_;
+
+/**
+ * Broadcasts a command instance to the listening cluster. This
+ * function must be called on the root whenever the cluster is in
+ * listening mode and errs otherwise.
+ *
+ * @param command shared pointer to an instance of a command class
+ * derived from mpi_command
+ * @return A unique_lock instance locking the mpi_cluster
+ */
```

Comment:
[optional]

Doc that `command` is intentionally being copied---it appears to be how pointer sharing works.

Reply:
hmm... actually we can make this a reference here, I think. I don't see a need to copy it. This is a shared_ptr as this is the only way to broadcast different command implementations through an instance of a virtual class. However, the shared_ptr must not be copied here. Let me switch that to a reference. that should work here just as well.